src/backend/distributed/shardsplit/shardsplit_logical_replication.c

```diff
@@ -0,0 +1,734 @@
+{
+	ShardSplitSubscriberMetadata *shardSplitPubSubMetadata = NULL;
+	foreach_ptr(shardSplitPubSubMetadata, shardSplitPubSubMetadataList)
+	{
+		Bitmapset *tableOwnerIds = NULL;
+		tableOwnerIds = bms_add_member(tableOwnerIds,
+									   shardSplitPubSubMetadata->tableOwnerId);
+		WaitForRelationSubscriptionsBecomeReady(
+			shardSplitPubSubMetadata->targetNodeConnection, tableOwnerIds,
+			SHARD_SPLIT_SUBSCRIPTION_PREFIX);
+	}
+}
+
+
+/*
+ * WaitForShardSplitRelationSubscriptionsToBeCaughtUp waits until subscriptions are caught up till
+ * the source LSN. This method invokes 'WaitForShardSubscriptionToCatchUp' for each subscription.
+ */
+void
+WaitForShardSplitRelationSubscriptionsToBeCaughtUp(XLogRecPtr sourcePosition,
+												   List *shardSplitPubSubMetadataList)
+{
+	ShardSplitSubscriberMetadata *shardSplitPubSubMetadata = NULL;
+	foreach_ptr(shardSplitPubSubMetadata, shardSplitPubSubMetadataList)
+	{
+		Bitmapset *tableOwnerIds = NULL;
+		tableOwnerIds = bms_add_member(tableOwnerIds,
+									   shardSplitPubSubMetadata->tableOwnerId);
+
+		WaitForShardSubscriptionToCatchUp(shardSplitPubSubMetadata->targetNodeConnection,
+										  sourcePosition,
+										  tableOwnerIds,
+										  SHARD_SPLIT_SUBSCRIPTION_PREFIX);
+	}
+}
+
+
+char *
+DropExistingIfAnyAndCreateTemplateReplicationSlot(ShardInterval *shardIntervalToSplit,
+												  MultiConnection *sourceConnection)
+{
+	/*
+	 * To ensure SPLIT is idempotent drop any existing slot from
+	 * previous failed operation.
+	 */
+	StringInfo dropReplicationSlotCommand = makeStringInfo();
+	appendStringInfo(dropReplicationSlotCommand, "SELECT pg_drop_replication_slot('%s')",
+					 ShardSplitTemplateReplicationSlotName(
+						 shardIntervalToSplit->shardId));
+
+	/* The Drop command can fail so ignore the response / result and proceed anyways */
+	PGresult *result = NULL;
+	int response = ExecuteOptionalRemoteCommand(sourceConnection,
+												dropReplicationSlotCommand->data,
+												&result);
+
+	PQclear(result);
+	ForgetResults(sourceConnection);
+
+	/*
+	 * Note: Temporary slot are only live during the session's lifetime causing them to be dropped when the session ends.
+	 */
+	StringInfo createReplicationSlotCommand = makeStringInfo();
+
+	appendStringInfo(createReplicationSlotCommand,
+					 "CREATE_REPLICATION_SLOT %s LOGICAL citus EXPORT_SNAPSHOT;",
+					 ShardSplitTemplateReplicationSlotName(
+						 shardIntervalToSplit->shardId));
+
+	response = ExecuteOptionalRemoteCommand(sourceConnection,
+											createReplicationSlotCommand->data, &result);
+
+	if (response != RESPONSE_OKAY || !IsResponseOK(result) || PQntuples(result) != 1)
+	{
+		ReportResultError(sourceConnection, result, ERROR);
+	}
+
+	/*'snapshot_name' is second column where index starts from zero.
+	 * We're using the pstrdup to copy the data into the current memory context */
+	char *snapShotName = pstrdup(PQgetvalue(result, 0, 2 /* columIndex */));
```

Comment:
We should probably use read with column name API instead, make it more readable.

Reply:
Are you saying we add a wrapper method over it? I see this pattern is common everywhere in CItus.